### PR TITLE
Implement contact form

### PR DESCRIPTION
**Pull request by timEulitz**
https://phabricator.wikimedia.org/T224080

### CODE REVIEW

File: skins/laika/src/components/pages/Contact.vue [hunk 1]
```diff
@@ -0,0 +1,202 @@
+<template>
+	<div class="column is-full contact-form">
+		<h1 class="title">{{ $t( 'contact_form_title' ) }}</h1>
+		<span class="help is-danger has-padding-bottom-18" v-if="contactData.errors">{{ $t('contact_form_error') }}</span>
+		<form method="post" action="/contact/get-in-touch" @submit="submit" id="laika-contact">
+			<fieldset>
+				<div>
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: unnecessary `div`

File: skins/laika/src/components/pages/Contact.vue [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+					</label>
+					<b-field>
+						<b-input type="text"
+								id="firstname"
+								name="firstname"
+								:placeholder="$t( 'contact_form_firstname_placeholder' )"
+								v-model="formData.name.value">
+						</b-input>
+					</b-field>
+					<span v-if="formData.name.validity === Validity.INVALID" class="help is-danger">{{ $t( 'contact_form_firstname_error' ) }}</span>
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: The logic in all the `v-if`s can be transferred into `data` or a computed property. Kinda like what we did in `Address.vue`.

File: skins/laika/src/components/pages/Contact.vue [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		return {
+			formData: {
+				name: {
+					name: 'name',
+					value: '',
+					pattern: '^.+$',
+					optionalField: true,
+					validity: Validity.VALID,
+				},
+				surname: {
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: in the name of consistency let's have `firstName` and `lastName`

File: skins/laika/src/components/pages/Contact.vue [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+				let field = this.$data.formData[ fieldName ];
+				field.validity = Helper.inputIsValid( field.value, field.pattern, field.optionalField );
+				if ( field.validity !== Validity.VALID ) {
+					isValid = false;
+				}
+			} );
+			if ( isValid ) {
+				return true;
+			}
+			event.preventDefault();
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: There are event modifiers in Vue which you can use directly in the template. 
In this case you can have `<form v-on:submit.prevent>` and remove this line from the ts.

File: skins/laika/src/components/pages/Contact.vue [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+		},
+	},
+	computed: {
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: not a blocker but can you please move `computed` after `props`

File: skins/laika/src/store/util.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import { Validity } from '@/view_models/Validity';
-import { PostData, FormData } from '@/view_models/Address';
+import { FormData, PostData } from '@/view_models/Address';
```
Comment (tzhelyazkova) [inline skins/laika/src/store/util.ts]: ? :laughing:

File: skins/laika/src/store/util.ts [hunk 7]
```diff
@@ -11,12 +12,21 @@ export const action = buildActionOrMutationName;
...
 export const Helper = {
-	inputIsValid: ( value: string, pattern: string ) => {
+	inputIsValid: ( value: string, pattern: string, isOptional?: boolean ) => {
+		if ( isOptional && Helper.isEmpty( value ) ) {
+			return Validity.VALID;
+		}
 		if ( pattern === null ) {
 			return value !== '' ? Validity.VALID : Validity.INVALID;
 		}
+		if ( Helper.isEmpty( value ) ) {
```
Comment (tzhelyazkova) [inline skins/laika/src/store/util.ts]: uhm, isn't this just `isEmpty( value )`

File: skins/laika/src/store/util.ts [hunk 8]
```diff
@@ -11,12 +12,21 @@ export const action = buildActionOrMutationName;
...
+			return Validity.INVALID;
+		}
 		return new RegExp( pattern ).test( value ) ? Validity.VALID : Validity.INVALID;
 	},
+	isEmpty( value: string ) {
+		return value === undefined || value === null || value === '';
```
Comment (tzhelyazkova) [inline skins/laika/src/store/util.ts]: With TypeScript that's technically impossible. We're always passing a string, so all the check should do is compare the value to an empty string.

File: skins/laika/src/view_models/Contact.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { Validity } from '@/view_models/Validity';
+
+export interface AddressValidity {
+    [key: string]: boolean
+}
+
+export interface InputField {
+    name: string,
+    value: string | null,
```
Comment (tzhelyazkova) [inline skins/laika/src/view_models/Contact.ts]: when is `value` null? According to the data object it's initially an empty string.

File: skins/laika/src/view_models/Contact.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { Validity } from '@/view_models/Validity';
+
+export interface AddressValidity {
```
Comment (tzhelyazkova) [inline skins/laika/src/view_models/Contact.ts]: I didn't see this used anywhere.

File: skins/laika/src/components/pages/Contact.vue [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+				</div>
+				<div class="has-margin-top-18">
+					<label for="category" class="subtitle">{{ $t( 'contact_form_topic_placeholder' ) }}</label>
+					<b-select
+						class="is-form-input"
+						v-model="formData.topic.value"
+						id="category"
+						name="category"
+						:placeholder="$t( 'contact_form_topic_placeholder' )"
+						@input="$emit('field-changed', 'title')">
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: probably a copy paste mistake?

File: skins/laika/src/components/pages/Contact.vue (hunk 4 above)
Comment (timEulitz) [reply skins/laika/src/components/pages/Contact.vue]: Cool :snowflake:

File: skins/laika/src/store/util.ts (hunk 6 above)
Comment (timEulitz) [reply skins/laika/src/store/util.ts]: I blame PHPStorm for autistically re-organizing that :man_shrugging:

File: skins/laika/src/view_models/Contact.ts (hunk 9 above)
Comment (timEulitz) [reply skins/laika/src/view_models/Contact.ts]: The drop down in its default state is `null` (empty string and undefined cause it to bug out) so I added `null` here. If we're putting on our fancy pants we can also create a new interface for just the subject dropdown but I figured this was okay and did not want to complicate things.

File: skins/laika/src/store/util.ts (hunk 8 above)
Comment (timEulitz) [reply skins/laika/src/store/util.ts]: I think it allows undefined and null to be passed anyway, that's why I had to add it. The subject drop down can be in a null state and I added undefined just to be sure.

File: skins/laika/src/store/util.ts (hunk 7 above)
Comment (timEulitz) [reply skins/laika/src/store/util.ts]: Typescript.

Comment (tzhelyazkova) [reply skins/laika/src/store/util.ts]: then it must be `this.isEmpty`, referencing its own name is just weird :D

File: skins/laika/src/store/util.ts (hunk 8 above)
Comment (tzhelyazkova) [reply skins/laika/src/store/util.ts]: We have strict mode set to true in the TS compiler options, so strict null checks is also on -> you can't pass undefined and null to a string.

Comment (tzhelyazkova) [reply skins/laika/src/store/util.ts]: If you actually want value here to be a string or null, we need a union type.

File: skins/laika/src/components/pages/Contact.vue (hunk 1 above)
Comment (timEulitz) [reply skins/laika/src/components/pages/Contact.vue]: All the other input elements and their labels / errors have this `div` so it's there for consistency's sake I would argue.

Comment (tzhelyazkova) [reply skins/laika/src/components/pages/Contact.vue]: It literally  doesn't do anything :D

File: skins/laika/src/components/pages/Contact.vue [hunk 12]
```diff
@@ -3,7 +3,7 @@
 		<h1 class="title">{{ $t( 'contact_form_title' ) }}</h1>
 		<span class="help is-danger has-padding-bottom-18" v-if="contactData.errors">{{ $t('contact_form_error') }}</span>
 		<span class="help is-danger has-padding-bottom-18" v-for="error in contactData.errors">{{ $t( error ) }}</span>
-		<form method="post" action="/contact/get-in-touch" @submit="submit" id="laika-contact">
+		<form method="post" action="/contact/get-in-touch" @submit="submit" v-on:submit.prevent id="laika-contact">
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: `@submit="submit"` is the same as `v-on:submit`

File: skins/laika/src/components/pages/Contact.vue (hunk 11 above)
Comment (tzhelyazkova) [reply skins/laika/src/components/pages/Contact.vue]: this is still here

File: skins/laika/src/components/pages/Contact.vue [hunk 13]
```diff
@@ -176,8 +176,15 @@ export default Vue.extend( {
...
+		Validity: {
+			get() {
+				return Validity;
+			},
+		},
+	},
 	methods: {
-		submit( event: Event ) {
+		submit() {
 			let isValid = true;
```
Comment (tzhelyazkova) [inline skins/laika/src/components/pages/Contact.vue]: what does this flag do? :thinking:

File: skins/laika/src/view_models/Contact.ts (hunk 10 above)
Comment (timEulitz) [reply skins/laika/src/view_models/Contact.ts]: Copy-pasta :spaghetti:

File: skins/laika/src/components/pages/Contact.vue (hunk 11 above)
Comment (timEulitz) [reply skins/laika/src/components/pages/Contact.vue]: Blagodarya!

File: skins/laika/src/components/pages/Contact.vue (hunk 13 above)
Comment (timEulitz) [reply skins/laika/src/components/pages/Contact.vue]: It's needed too only submit the form conditionally. I have updated that part of the code now to not use the event.

**Pull request merged**